src/TableFunctions/TableFunctionS3Cluster.cpp

```diff
@@ -45,102 +46,104 @@ void TableFunctionS3Cluster::parseArguments(const ASTPtr & ast_function, Context
         " - cluster, url, format, structure, compression_method\n" \
+        " - cluster, url, access_key_id, secret_access_key, format\n"
         " - cluster, url, access_key_id, secret_access_key, format, structure\n" \
         " - cluster, url, access_key_id, secret_access_key, format, structure, compression_method",
         getName());
 
-    if (args.size() < 4 || args.size() > 7)
+    if (args.size() < 2 || args.size() > 7)
         throw Exception(message, ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
 
-    for (auto & arg : args)
-        arg = evaluateConstantExpressionOrIdentifierAsLiteral(arg, context);
-
     /// This arguments are always the first
-    cluster_name = args[0]->as<ASTLiteral &>().value.safeGet<String>();
-    filename = args[1]->as<ASTLiteral &>().value.safeGet<String>();
-
-    /// Size -> argument indexes
-    static auto size_to_args = std::map<size_t, std::map<String, size_t>>
-    {
-        {4, {{"format", 2}, {"structure", 3}}},
-        {5, {{"format", 2}, {"structure", 3}, {"compression_method", 4}}},
-        {6, {{"access_key_id", 2}, {"secret_access_key", 3}, {"format", 4}, {"structure", 5}}},
-        {7, {{"access_key_id", 2}, {"secret_access_key", 3}, {"format", 4}, {"structure", 5}, {"compression_method", 6}}}
-    };
-
-    auto & args_to_idx = size_to_args[args.size()];
-
-    if (args_to_idx.contains("format"))
-        format = args[args_to_idx["format"]]->as<ASTLiteral &>().value.safeGet<String>();
-
-    if (args_to_idx.contains("structure"))
-        structure = args[args_to_idx["structure"]]->as<ASTLiteral &>().value.safeGet<String>();
+    configuration.cluster_name = args[0]->as<ASTLiteral &>().value.safeGet<String>();
 
-    if (args_to_idx.contains("compression_method"))
-        compression_method = args[args_to_idx["compression_method"]]->as<ASTLiteral &>().value.safeGet<String>();
+    if (!context->tryGetCluster(configuration.cluster_name))
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Cannot get {} cluster", configuration.cluster_name);
 
-    if (args_to_idx.contains("access_key_id"))
-        access_key_id = args[args_to_idx["access_key_id"]]->as<ASTLiteral &>().value.safeGet<String>();
+    /// Just cut the first arg (cluster_name) and try to parse s3 table function arguments as is
+    ASTs clipped_args;
+    clipped_args.reserve(args.size());
+    std::copy(args.begin() + 1, args.end(), std::back_inserter(clipped_args));
 
-    if (args_to_idx.contains("secret_access_key"))
-        secret_access_key = args[args_to_idx["secret_access_key"]]->as<ASTLiteral &>().value.safeGet<String>();
+    /// StorageS3ClusterConfiguration inherints from StorageS3Configuration, so it is safe to upcast it.
+    TableFunctionS3::parseArgumentsImpl(message, clipped_args, context, static_cast<StorageS3Configuration & >(configuration));
 }
 
 
 ColumnsDescription TableFunctionS3Cluster::getActualTableStructure(ContextPtr context) const
```

Comment:
We should also add schema inference here, look at the `TableFunctionS3::getActualTableStructure`. It's needed to be able to run queries like 'desc s3Cluster(...)' that will call `getActualTableStructure` without storage creation.